src/Compilers/CSharp/Portable/Symbols/Source/SourceLocalSymbol.cs

```diff
@@ -580,5 +598,58 @@ protected override TypeSymbol InferTypeOfVarVariable(DiagnosticBag diagnostics)
                 }
             }
         }
+
+        /// <summary>
+        /// Symbol for a deconstruction local that might require type inference.
+        /// For instance, local `x` in `var(x, y) = ...` or `(var x, int y) = ...`.
+        /// </summary>
+        private class PossiblyImplicitlyTypedDeconstructionLocalSymbol : SourceLocalSymbol
+        {
+            public PossiblyImplicitlyTypedDeconstructionLocalSymbol(
+                Symbol containingSymbol,
+                Binder binder,
+                TypeSyntax typeSyntax,
+                SyntaxToken identifierToken,
+                LocalDeclarationKind declarationKind)
+            : base(containingSymbol, binder, RefKind.None, typeSyntax, identifierToken, declarationKind)
+            {
+            }
+
+            protected override TypeSymbol InferTypeOfVarVariable(DiagnosticBag diagnostics)
+            {
+                // Try binding immediately enclosing deconstruction-declaration (the top-level VariableDeclaration), this should force the inference.
+
+                var topLevelVariableDeclaration = (VariableDeclarationSyntax)IdentifierToken.
+                                                                    Parent. // VariableDeclaratorSyntax
+                                                                    Parent; // VariableDeclarationSyntax
+                while (true)
+                {
+                    var parent = topLevelVariableDeclaration.Parent;
+                    if (parent.Kind() != SyntaxKind.VariableDeconstructionDeclarator) { break; }
+
+                    var grandParent = parent.Parent;
+                    if (grandParent.Kind() != SyntaxKind.VariableDeclaration) { break; }
+
+                    topLevelVariableDeclaration = (VariableDeclarationSyntax)grandParent;
+                }
+
+                Debug.Assert(topLevelVariableDeclaration.Deconstruction != null && topLevelVariableDeclaration.Deconstruction.Value != null);
+
+                var statement = topLevelVariableDeclaration.Parent;
+                TypeSymbol result;
+                switch (statement.Kind())
+                {
+                    case SyntaxKind.LocalDeclarationStatement:
+                        this.binder.BindDeconstructionDeclarationStatementParts((LocalDeclarationStatementSyntax)statement, diagnostics);
```

Comment:
This might be not the right binder to use, this is the binder that owns the scope for this local, but the local declaration statement has its own scope for pattern and out var locals. That binder should be in the chain at this point. Try using out vars and pattern vars in the expression part of the declaration. It looks like we have similar problem with LocalWithInitializer and other similar C# 6 types. I suggest to not try to be smart here and simply call the method that binds the whole declaration statement, it should locate the right binder for the job.